src/nvim/tui/tui.c

```diff
@@ -1007,28 +996,33 @@ static void tui_mode_change(UI *ui, String mode, Integer mode_idx)
   data->showing_mode = (ModeShape)mode_idx;
 }
 
-static void tui_grid_scroll(UI *ui, Integer g, Integer top, Integer bot,
-                            Integer left, Integer right,
+static void tui_grid_scroll(UI *ui, Integer g, Integer startrow, Integer endrow,
+                            Integer startcol, Integer endcol,
                             Integer rows, Integer cols)
 {
+  int top = (int)startrow, bot = (int)endrow-1;
+  int left = (int)startcol, right = (int)endcol-1;
   TUIData *data = ui->data;
   UGrid *grid = &data->grid;
-  ugrid_set_scroll_region(&data->grid, (int)top, (int)bot-1,
-                          (int)left, (int)right-1);
 
-  data->scroll_region_is_full_screen =
-    left == 0 && right == ui->width
-    && top == 0 && bot == ui->height;
+  bool fullwidth = left == 0 && right == ui->width-1;
+  data->scroll_region_is_full_screen = fullwidth
+        && top == 0 && bot == ui->height-1;
 
   int clear_top, clear_bot;
-  ugrid_scroll(grid, (int)rows, &clear_top, &clear_bot);
-
-  if (can_use_scroll(ui)) {
+  ugrid_scroll(grid, top, bot, left, right, (int)rows,
+               &clear_top, &clear_bot);
+
+  if (data->scroll_region_is_full_screen
+      || (data->can_change_scroll_region
+          && ((left == 0 && right == ui->width - 1)
+              || data->can_set_lr_margin
+              || data->can_set_left_right_margin))) {
```

Comment:
if not in a function, would still be worth assigning this to a local `can_use_scroll` boolean.